src/js/components/PlacementConstraintsFrameworkAdapter.js

```diff
@@ -0,0 +1,88 @@
+import React, { Component } from "react";
+import Batch from "#SRC/js/structs/Batch";
+import PlacementConstraintsPartial
+  from "#SRC/js/components/PlacementConstraintsPartial";
+import BatchContainer from "#SRC/js/components/BatchContainer";
+import {
+  JSONReducer,
+  JSONParser
+} from "#PLUGINS/services/src/js/reducers/serviceForm/JSONReducers/Constraints";
+import { combineReducers } from "#SRC/js/utils/ReducerUtil";
+import {
+  FormReducer
+} from "#PLUGINS/services/src/js/reducers/serviceForm/FormReducers/Constraints";
+import CreateServiceModalFormUtil
+  from "#PLUGINS/services/src/js/utils/CreateServiceModalFormUtil";
+
+const jsonReducer = combineReducers({ constraints: JSONReducer });
+
+export default class PlacementConstraintsFrameworkAdapter extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      batch: this.generateBatchFromInput()
+    };
+
+    this.handleBatchChange = this.handleBatchChange.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.data !== this.props.data) {
+      this.setState({
+        batch: this.generateBatchFromInput(nextProps)
+      });
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    const propsChanged = nextProps.data !== this.props.data;
+    const stateChanged = nextState.batch !== this.state.batch;
+
+    return propsChanged || stateChanged;
+  }
+
+  generateBatchFromInput(props = this.props) {
+    const { data } = props;
+
+    const parsedInput = CreateServiceModalFormUtil.stripEmptyProperties(
+      JSON.parse(data)
```

Comment:
We should `try catch` this in the case when exception in parsing the JSON data. Then I think just return the previous batch, to prevent going into a weird state:

```
    let parsedInput;
    try {
      parsedInput = CreateServiceModalFormUtil.stripEmptyProperties(
        JSON.parse(data)
      );
    } catch (e) {
      return this.state.batch;
    }
```

To reproduce, just delete one of the brackets in the JSON editor "placement_constraints

Reply:
Good one! Thank you!